### PR TITLE
Further optimizations for String.Join, String.Concat

### PR DESCRIPTION
**Pull request by jamesqo**
Changes:

- We call the `IEnumerable<string>` based overload in the generic overloads, if `typeof(T) == typeof(string)`. This can happen practically if someone has a `GenericMethod<T>` that calls Concat/Join, and then someone calls that method with T == string.

- Optimize more Join/Concat overloads for length 0-and-1 collections by not allocating anything.

- Cleanup some of the code.

cc @jkotas, @bbowyersmyth 

Tests are dotnet/corefx#10966

### CODE REVIEW

File: src/mscorlib/src/System/String.cs [hunk 1]
```diff
@@ -107,6 +118,15 @@ public static String Join<T>(String separator, IEnumerable<T> values)
             Contract.Ensures(Contract.Result<String>() != null);
             Contract.EndContractBlock();
 
+            // This can be called with T == string if someone
+            // has a GenericMethod<T> which calls here, and
+            // then someone calls that with string
+            if (typeof(T) == typeof(string))
```
Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: I do not think that this optimization is worth the extra code.

Comment (jamesqo) [inline src/mscorlib/src/System/String.cs]: @jkotas Sure, removed the optimizations.

**Comment by jasonwilliams200OK** [comment]
@jamesqo, this will produce merge conflicts with #2945, which is similar to #895.

**Comment by jamesqo** [comment]
@jasonwilliams200OK I have opened up a pull request in your fork which I believe should remove the merge conflicts with this PR.

**Comment by jamesqo** [comment]
@jkotas Is this OK to be merged? I will port the changes to the CoreRT repo if it's ready.

**Comment by jkotas** [comment]
LGTM. Thanks!

**Pull request merged**